src/UserCodeRunner.ts

```diff
@@ -1,482 +1,649 @@
-    if (diagnostic.file) {
-      sourceErrors.push(UserCodeTypeError.new(diagnostic, tsFileCache));
-    }
-  }
-
-  if (sourceErrors.length > 0) {
-    return Result.Err(sourceErrors);
-  }
-
-  // Put args and result into context
-  context.args = args;
-  context.result = undefined;
-
-  // Create modules for VM
-  const moduleCache = new Map<string, vm.Module>();
-  for (const jsFile of jsFileCache.values()) {
-    if (jsFile.fileName.endsWith('.js')) {
-      moduleCache.set(jsFile.fileName, new vm.SourceTextModule(jsFile.text, {
-        identifier: jsFile.fileName,
-        context,
-      }));
-    }
-  }
-  const harnessModule = moduleCache.get(`${EXECUTION_HARNESS_FILENAME}.js`)!;
-  await harnessModule.link((specifier) => {
-    if (moduleCache.has(specifier + '.js')) {
-      return moduleCache.get(specifier + '.js')!;
-    }
-    throw new Error(`Unable to resolve dependency: ${specifier}`);
-  });
-
-  try {
-    await harnessModule.evaluate({
-      timeout
-    });
-    return Result.Ok(context.result);
-  } catch (error: any) {
-    return Result.Err([UserCodeRuntimeError.new(error as Error, sourceMap, tsFileCache)]);
-  }
+      result = defaultExport(...args);
+    `;
+		const executionSourceFile = ts.createSourceFile(
+			EXECUTION_HARNESS_FILENAME,
+			executionCode,
+			ts.ScriptTarget.ESNext,
+			undefined,
+			ts.ScriptKind.TS,
+		);
+
+		const tsFileMap = new Map<string, ts.SourceFile>();
+
+		tsFileMap.set(`${USER_FILE_ALIAS}.ts`, userSourceFile);
+		tsFileMap.set(`${EXECUTION_HARNESS_FILENAME}.ts`, executionSourceFile);
+
+		for (const additionalSourceFile of additionalSourceFiles) {
+			tsFileMap.set(additionalSourceFile.fileName, additionalSourceFile);
+		}
+
+		const jsFileMap = new Map<string, ts.SourceFile>();
+
+		const defaultCompilerHost = ts.createCompilerHost({});
+		const customCompilerHost: ts.CompilerHost = {
+			...defaultCompilerHost,
+			getCurrentDirectory(): string {
+				return '';
+			},
+			getSourceFile: (fileName, languageVersion) => {
+				if (tsFileMap.has(fileName)) {
+					return tsFileMap.get(fileName);
+				} else if (fileName.includes('typescript/lib')) {
+					return defaultCompilerHost.getSourceFile(fileName, languageVersion);
+				}
+				return undefined;
+			},
+			writeFile: (filename, data) => {
+				jsFileMap.set(
+					filename,
+					ts.createSourceFile(filename, data, ts.ScriptTarget.ESNext, undefined, ts.ScriptKind.JS),
+				);
+			},
+			readFile(fileName: string): string | undefined {
+				if (tsFileMap.has(fileName)) {
+					return tsFileMap.get(fileName)!.text;
+				}
+				return defaultCompilerHost.readFile(fileName);
+			},
+			fileExists(fileName: string): boolean {
+				return tsFileMap.has(path.basename(fileName));
+			},
+		};
+
+		const program = ts.createProgram(
+			[...additionalSourceFiles.map(f => f.fileName), EXECUTION_HARNESS_FILENAME],
+			{
+				target: ts.ScriptTarget.ESNext,
+				module: ts.ModuleKind.ES2022,
```

Comment:
Likely doesn't matter but on our `feature/aerie-1741-ts-activity-instances` aerie branch we're using `ES2021`. Should that be changed to `ES2022`?

Reply:
We can't target ES2021 because it isn't available yet? I tried when I created the commanding server but npm yelled at me.